### PR TITLE
Microsoft.StorageCache 2024-03-01 API

### PR DESCRIPTION
**Pull request by rebecca337**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [X] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [X] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [X] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (raosuhas) [inline specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:495]: why not make this operation long running instead of asking the user to look at some other object for status ?

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (raosuhas) [inline specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:1183]: >importId [](http://example.com/codeflow?start=9&length=8)

what is this importid ? does it refer to some other well known id ? Can you also update the description ?

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (raosuhas) [inline specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:515]: >cancelImport [](http://example.com/codeflow?start=140&length=12)

ARM much recommends modelling operations using CRUD : PUT\GET\DELETE. since most tools written on top of ARM only work with these for example ARM templates \ ARM policy \ ARG etc 


Consider if this could be better modelled as a PUT on an ImportJob, which is a long running operation and gives you the status at the end

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (brpanask) [reply specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:495]: The api imports information from large storage accounts that could take days.   The job will complete and go away when it's done.  

If this is a resource, then the customer has to delete and cleanup old ones.  There are no settings to change other than to cancel it to start a new one.   The status is an array in case in the future we would have the ability to have parallel import jobs, so the provided key was added as the x-ms-identifier instead of trying to use the array of import prefixes as a potential key.  Right now there is only one, but the idea was that it would be a breaking change to make it an array, so we went with an array to start.    

What is the limit how long an async op can run?  This could take a while depending on how large the dataset they want to import is?

Should we use a sub resource for jobs that would just go away?  If it's a sub resource we can't auto delete them, can we?
At some point I think we would want to block additional imports if they haven't cleaned out the old ones.

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (rebecca337) [reply specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:1183]: @raosuhas answered by Brian above. It serves as a key for the import for a future case where we may have than one

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (rebecca337) [reply specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:515]: @raosuhas Brian explained above why we chose POST but to expand, the import job will run to completion and then get removed without further interaction needed from the customer. We see an import/cancelImport as an attribute for the AMLFS resource. If this were a sub-resource, we could no longer expose the status/details in a single GET AMLFS, which would not be optimal for our customers.

The import/cancelImport are actions on the amlfs resource that import data to the amlfs cluster. These actions are similar to actions on other resources like ["e.g. getting the thumbnail of an image or rebooting a VM"](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#performing-an-action)

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (visingla-ms) [reply specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:495]: My understanding from above is:

When you define it as an action named "Import", and it returns 200 response, it could be interpreted as that "Import has completed".

How about you model it an Post Action named "startImportJob"? Then, as soon as your backend is able to start the job successfully, you return 200 response to user. The job can continue to work and complete the import behind the scenes. For user's tracking, the 200 response of "startImportJob" gives the user an "operationId", using which the user can track the status of import operation (e.g. 10% complete or fully completed etc.).

Even better will be model it as PUT "/importJobs/{jobName}", and user can provide some name to job and does a PUT to the job. When the job is created successfully, PUT returns with 201 response, along with an operationId for user to track the status of import.

The importJobs can be a proxy resource, with only GET, PUT defined. If delete is not required, don't implement DELETE API.

What do you think of these suggestions?

Tagging @raosuhas for his view points also.

Comment (brpanask) [reply specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:495]: @visingla-ms,

Here's some additional information.

There is an import status being added in the amlfilesystem resource that has percent complete, detailed errors, start and completion time, and state. It's all readonly and can be shown in the GET.  

The previous versions did an import as part of PUT supplying the container to import from and any optional filesystem paths to get within the container.  That is still available and kicks off the import as part of the initial PUT to import data into the cluster.   This change allows the customer to import additional information after the initial PUT which wasn't possible before.  For backward compatibility we still need to import during the initial creation.

In the current version, the amlfilesystem PUT waits on the import to complete during provisioning of the cluster.  This can make just building the cluster take a long time because it included the import.  Now the import would only kicked off from PUT and the status is provided to watch for its completion and for status of additional imports.

The importStatus is an array instead of a single object to handle the future case of potentially handling parallel imports from different filesystem paths.  We currently don't support that so there would only be 1 at a time.   The key is the array of filesystem paths to pull from, which is complex to use as a key.  The unique id was used to key the potential parallel imports for the importStatus array if that becomes available.

1.  Add LRO details to the post operations and update the name.
If we use Post Action startImportJob, its similar to what we have.
The long running operation id can be returned with status.
Would you expect the detailed importStatus to go away?  It covers the status of the import kicked off during the initial PUT amlfilesystem.  It also has all of the details and can always be found by using list or get on your amlfilesystems in case you lose the async op id given that these may take hours to days.

2. If we use the proxy resource with only GET/PUT,
The resource isn't tied in the swagger spec to amlfilesystem directly.  Its floating in with other things like operations and async operations apis making it more difficult to find in the docs and link to the amlfilesystem object that the job is for.
The customer has to monitor that resource for status using its GET.   If this is a separate resource then they can't just see how the import is going on this cluster with GET on amlfilesystem.  Instead they would have to monitor GET on another resource.   
Would the importStatus we have in the amlfilesystem go away or also still exist?
Would we need a collection GET to see them all since its a separate resource?
What if they lose the name/id of this job over the hours to days it takes to complete?

3. As a sub resource of amlfilesystem,
If this becomes a sub resource, since this is an operation, not a setting, it should go away at some point so they don't build up. 
Can these sub resource's be auto deleted at some point after they reach a terminal state of complete, cancelled, failed?  
If not, do we block creation of new ones while a specific amount exists forcing them to delete them?

This would be good in that it would make the status array from amlfilesystem GET go away and the list/get on the sub resources would show the details and provide a sort of history up until the auto deletion point.
It could provide the async behavior so that the import can be tracked from the PUT.
It would allow imports to work with ARM templates and other things that work with resources better than POST operations.1

Can an initial sub resource be provided in the amlfilesystem creation PUT to start the import of the data into the cluster?  The current api supports that so we would have to do that in some way.

**Comment by rebecca337** [comment]
azp run

File: specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json
Comment (rkmanda) [inline specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:596]: >description [](http://example.com/codeflow?start=13&length=11)

Would u like to call out any restrictions for this PUT operation? Are you going with the approach of adding a service defined resource instance to this collection which will be readOnly?

Comment (rkmanda) [reply specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json:596]: Meaning would u reject the put for that specific named instance - maybe called initialImportJob?

**Comment by kazrael2119** [comment]
please fix duplicate schemas : ErrorResponse, Resource, TrackedResource.
please check the reference version from common-type

**Comment by rebecca337** [comment]
@kazrael2119 The 6 commits from today are my attempts to fix the duplicate schema errors in the SDK pipeline and sdk go. 
The ArmSignOff tag and Approved-SdkBreakingChange-Python tag were removed. Could they be added back?

Finally, I cannot fix the duplicate schema error without causing another breaking change error. The "ErrorResponse" schema added to this PR is from the **v5 common types**. In storagecache.json we have a **pre-existing "ErrorResponse" schema** that we used for older API versions. I cannot remove that one without causing breaking changes, but I have to use the v5 common type schema for the new api or else I get other errors... 
Please advise what the best course is.

**Comment by kazrael2119** [comment]
> @kazrael2119 The 6 commits from today are my attempts to fix the duplicate schema errors in the SDK pipeline and sdk go. The ArmSignOff tag and Approved-SdkBreakingChange-Python tag were removed. Could they be added back?
> 
> Finally, I cannot fix the duplicate schema error without causing another breaking change error. The "ErrorResponse" schema added to this PR is from the **v5 common types**. In storagecache.json we have a **pre-existing "ErrorResponse" schema** that we used for older API versions. I cannot remove that one without causing breaking changes, but I have to use the v5 common type schema for the new api or else I get other errors... Please advise what the best course is.

@rebecca337 sorry I have no idea, maybe you have to ask BreakingChangeReviewBoard to get approval. Or we can't approve sdk breaking changes for JS and Go,

**Comment by ms-henglu** [comment]
There are a bunch of breaking changes reported on this PR. Please get a breaking changes approval from Azure Breaking Changes Reviewers <azbreakchangereview@microsoft.com> and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

**Comment by rebecca337** [comment]
Summary for next ARM reviewer: This PR was approved by ARM 2 weeks ago. After that, when trying to fix validation and pipeline errors, the ARM sign off tag was removed. The only changes since that review were schema/formatting changes to pass the checks. Also the Breaking Change reported is only for SDK breaking change, not API.

**Review by ramoka178 — APPROVED** [review]
Everything is Green. Approving to merge.

**Pull request merged**
